File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    /*
    s = "AABABBA", k = 1
    
    => ans = 4
    
    sliding window 
    */
 
    let left = 0;
    let right = 0;
    let map = new Map();
    let max = 0;
    
    while(right < s.length){
        let curChar = s[right];
        map.set(curChar, (map.get(curChar) || 0) + 1 );
        
        max = Math.max(map.get(curChar), max)
        
        if(right-left+1 - max > k){
            map.set(s[left], map.get(s[left]) -1);
            left++;
        }  right ++;
        
    } return right - left
}

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left = maxFreq = longest = 0
        d = {}
        
        for right in range(len(s)):
            d[s[right]] = d.get(s[right], 0 )+1
            maxFreq = max(maxFreq, d[s[right]])
            
            while right - left + 1 - maxFreq > k:
                d[s[left]] -= 1
                left += 1
                
            longest = max(longest, right-left+1)
            
        return longest